antenna/app.py

```diff
@@ -3,223 +3,211 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 import cgi
+import hashlib
 import io
 import json
 import logging
 import time
 import zlib
 
 from everett.manager import ConfigManager, ConfigOSEnv, parse_class
+from everett.component import ConfigOptions, RequiredConfigMixin
 import falcon
 
 from antenna.lib.datetimeutil import utc_now
 from antenna.lib.ooid import create_new_ooid
-from antenna.lib.storage import Storage
-from antenna.throttler import DISCARD, IGNORE
+from antenna.util import de_null
+
 
 logger = logging.getLogger('gunicorn.error')
 
 
-# FIXME: Figure out whether we need to implement this or nix it. For now, we're
-# just going to define it as a synonym.
-MemoryDumpsMapping = dict
+class BreakpadSubmitterResource(RequiredConfigMixin):
+    """Handles incoming breakpad crash reports and saves to S3"""
+    required_config = ConfigOptions()
+    required_config.add_option(
+        'dump_field', default='upload_file_minidump',
+        doc='the name of the field in the POST data for dumps'
+    )
+
+    required_config.add_option(
+        'dump_id_prefix', default='bp-',
+        doc='the crash type prefix'
+    )
 
+    required_config.add_option(
+        'crashstorage_class',
+        default='antenna.external.boto.crashstorage.BotoS3CrashStorage',
+        parser=parse_class,
+        doc='the class in charge of storing crashes'
+    )
 
-class BreakpadSubmitterResource(object):
     def __init__(self, config):
-        # the default name for the main dump
-        self.dump_field = config(
-            'dump_field', default='upload_file_minidump'
-        )
+        self.config = config.with_options(self)
+        self.crashstorage = self.config('crashstorage_class')(config)
 
-        # the prefix to return to the client in front of the OOID
-        self.dump_id_prefix = config(
-            'dump_id_prefix', default='bp-'
-        )
+    def process_fieldstorage(self, fs):
```

Comment:
TODO: Add more testing for this. I should write a bunch of multipart/form-data things and then parse them and verify the output. There are a few multipart libs out there which I can examine the tests for to see variations that should get tested.